snowfakery/data_generator_runtime.py

```diff
@@ -108,52 +108,17 @@ class Dependency(NamedTuple):
 )
 
 
-class SlotState(Enum):
-    """The current state of a NicknameSlot.
-
-    UNUSED=empty, ALLOCATED=referenced, CONSUMED=object generated"""
-
-    UNUSED = auto()
-    ALLOCATED = auto()
-    CONSUMED = auto()
-
-
-class NicknameSlot:
-    """A slot that represents a Nickname or Tablename"""
-
-    _tablename: str
-    id_manager: IdManager
-    allocated_id: int = None
-
-    def __init__(self, tablename, id_manager):
-        self._tablename = tablename
-        self.id_manager = id_manager
-
-    @property
-    def id(self):
-        "Get an id corresponding to this slot. Generate one if necessary."
-        if self.allocated_id is None:
-            self.allocated_id = self.id_manager.generate_id(self._tablename)
-        return self.allocated_id
-
-    def consume_slot(self):
-        "Mark a slot as filled by an object and return its id for use by the object."
-        rc = self.allocated_id
-        self.allocated_id = SlotState.CONSUMED
-        return rc
-
-    @property
-    def status(self):
-        "Is the slot empty/unreferenced, allocated/referenced or consumed/used"
-        if self.allocated_id is None:
-            return SlotState.UNUSED
-        elif isinstance(self.allocated_id, int):
-            return SlotState.ALLOCATED
-        elif self.allocated_id == SlotState.CONSUMED:
-            return SlotState.CONSUMED
-
-    def __repr__(self):
-        return f"<NicknameSlot {vars(self)}>"
+class Transients:
+    def __init__(self, nicknames_and_tables: Mapping[str, str], id_manager: IdManager):
+        self.nicknamed_objects = {}
+        self.last_seen_obj_by_table = {}
+        self.named_slots = {
+            name: NicknameSlot(table, id_manager)
+            for name, table in nicknames_and_tables.items()
+        }
+        self.orig_used_ids = defaultdict(
+            lambda: 1, {k: v + 1 for k, v in id_manager.last_used_ids.items()}
+        )
```

Comment:
This might warrant a comment -- took me a bit to figure out how this is being used.

Would it be worth changing the id manager to use 1-based indexing too?

Reply:
Why don't you explain it to me so I can comment it. :) (just kidding....)

Reply:
I think its much clearer in f3b2629 and the new naming (last_used_id versus first_new_id) probably clarifies why one should be 0-based and the other 1-based.

Reply:
@prescod yes, thanks, this is much clearer.